### PR TITLE
Refactor/add the rest of route generator refactor

### PR DESCRIPTION
**Pull request by eb-so**

### CODE REVIEW

File: packages/stacked_generator/test/router/route_class_generator_test.dart [hunk 1]
```diff
@@ -438,6 +438,18 @@ void main() {
 
         generateRouteAndExpectResult(routes, kRouterWithRetrunType);
       });
+      test('When adding returnType = <CupertinoRoute>', () {
+        final routes = [
+          MaterialRouteConfig(
+              name: 'loginView',
+              pathName: 'pathNamaw',
+              className: 'ebraClass',
```
Comment (Eimen2018) [inline packages/stacked_generator/test/router/route_class_generator_test.dart]: What is an ebraClass? 😅

File: packages/stacked_generator/lib/src/generators/router/router_class_generator.dart [hunk 2]
```diff
@@ -320,6 +310,24 @@ class RouterClassGenerator extends BaseGenerator {
     writeLine(');');
   }
 
+  RouteConfig _extractReturnType(RouteConfig r) {
+    final returnTypeContainsBiggerOperatorWithOneOfRouteNames = r.returnType !=
+            null &&
+        r.returnType!.contains('<') &&
+        r.returnType!.contains(
+            RegExp('CustomRoute|MaterialRoute|CupertinoRoute|AdaptiveRoute'));
```
Comment (Eimen2018) [inline packages/stacked_generator/lib/src/generators/router/router_class_generator.dart]: 🔥

File: packages/stacked_generator/lib/src/generators/router/route_config/route_config.dart [hunk 3]
```diff
@@ -60,6 +60,42 @@ class RouteConfig {
     return {...this.imports, ...guardsImports, ...paramertersImports};
   }
 
+  String registerArgs() {
+    StringBuffer stringBuffer = StringBuffer();
+
+    if (parameters.isNotEmpty == true) {
```
Comment (Eimen2018) [inline packages/stacked_generator/lib/src/generators/router/route_config/route_config.dart]: can't this be just `if(parameters.isNotEmpty)`?

File: packages/stacked_generator/lib/src/generators/router/route_config/route_config.dart [hunk 4]
```diff
@@ -60,6 +60,42 @@ class RouteConfig {
...
+      // if router has any required or positional params the argument class holder becomes required.
+      final nullOk =
+          !notQueryAndNotPath.any((p) => p.isRequired || p.isPositional);
+      // show an error page  if passed args are not the same as declared args
+
+      if (notQueryAndNotPath.isNotEmpty) {
```
Comment (Eimen2018) [inline packages/stacked_generator/lib/src/generators/router/route_config/route_config.dart:71]: ⬆️  Like this?

File: packages/stacked_generator/test/helpers/router_constants.dart [hunk 5]
```diff
@@ -1019,6 +1019,36 @@ RouteDef(RoutesClassB.loginView
         final _pagesMap = <Type, StackedRouteFactory>{
         
 
+ebraClass: (data) {
```
Comment (Eimen2018) [inline packages/stacked_generator/test/helpers/router_constants.dart]: ebraClass, this must be a signature or something 😂

**Review by Eimen2018 — COMMENTED** [review]
LGTM

File: packages/stacked_generator/test/router/route_class_generator_test.dart (hunk 1 above)
Comment (eb-so) [reply packages/stacked_generator/test/router/route_class_generator_test.dart]: Well, I didn't want to use stacked or filledstacks cause I needed a unique word, ebra is apprivation for ebrahim(me) 😆 
When you modify this class you can change it to AymanClass😜

File: packages/stacked_generator/lib/src/generators/router/route_config/route_config.dart (hunk 3 above)
Comment (eb-so) [reply packages/stacked_generator/lib/src/generators/router/route_config/route_config.dart]: Nice catch

File: packages/stacked_generator/lib/src/generators/router/route_config/route_config.dart (hunk 4 above)
Comment (eb-so) [reply packages/stacked_generator/lib/src/generators/router/route_config/route_config.dart:71]: Yes I fix this before but there is too many many of them like the above one(the code is so old)

File: packages/stacked_generator/test/router/route_class_generator_test.dart (hunk 1 above)
Comment (Eimen2018) [reply packages/stacked_generator/test/router/route_class_generator_test.dart]: @eb-so Nice 🔥 , but I think TestClass is appropriate 😂

File: packages/stacked_generator/lib/src/generators/router/generator/router_class_generator.dart [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        [..._rootRouterConfig.routes, ..._rootRouterConfig.subRouters]);
+    generateRouterByRouter(
+        _rootRouterConfig.routes,
+        _rootRouterConfig.routerClassName,
+        _rootRouterConfig.routesClassName,
+        _rootRouterConfig.generateNavigationHelper);
+    return stringBuffer.toString();
+  }
+
+  void generateRouterByRouter(
```
Comment (FilledStacks) [inline packages/stacked_generator/lib/src/generators/router/generator/router_class_generator.dart]: What does this function do? I'd like it to have a better name.

File: packages/stacked_generator/lib/src/generators/router/generator/route_generator_helper.dart [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+    writeLine('''
+     @override
+     List<RouteDef> get routes => _routes;
+     final _routes = <RouteDef>[
+     ''');
+    for (var route in routes) {
+      _generateRouteTemplates(route, routesClassName);
+    }
+    writeLine();
+    stringBuffer.write('];');
```
Comment (FilledStacks) [inline packages/stacked_generator/lib/src/generators/router/generator/route_generator_helper.dart]: Why do some lines write directly to the stringBuffer and others use the write and writeLine function? 

We should only use on style of writing.

File: packages/stacked_generator/lib/src/generators/router/generator/route_generator_helper.dart [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+
+    stringBuffer.write('};');
+
+    writeLine('}');
+  }
+
+  void _generateRouteTemplates(RouteConfig route, String routesClassName) {
+    writeLine();
+
+    stringBuffer.write("RouteDef(${routesClassName}.${route.templateName}");
```
Comment (FilledStacks) [inline packages/stacked_generator/lib/src/generators/router/generator/route_generator_helper.dart]: same goes for this section. Use one of the write/writeLine functions for this functionality.

File: packages/stacked_generator/test/logging/integration_test/samples/logger_disablereleasemode.dart [hunk 9]
```diff
@@ -1,6 +1,7 @@
 import 'package:stacked_core/stacked_core.dart';
 
 @StackedApp(
+  routes: [],
```
Comment (FilledStacks) [inline packages/stacked_generator/test/logging/integration_test/samples/logger_disablereleasemode.dart:4]: I don't see routes changed to required, but I just want to mention that it shouldn't be a required parameter.

**Review by FilledStacks — CHANGES_REQUESTED** [review]
Nice PR. I have a few small changes required

File: packages/stacked_generator/lib/src/generators/router/generator/router_class_generator.dart (hunk 6 above)
Comment (eb-so) [reply packages/stacked_generator/lib/src/generators/router/generator/router_class_generator.dart]: This function is created to generate the routes and their subroutes recursively

File: packages/stacked_generator/lib/src/generators/router/generator/route_generator_helper.dart (hunk 7 above)
Comment (eb-so) [reply packages/stacked_generator/lib/src/generators/router/generator/route_generator_helper.dart]: True I'll fix it

File: packages/stacked_generator/test/logging/integration_test/samples/logger_disablereleasemode.dart (hunk 9 above)
Comment (eb-so) [reply packages/stacked_generator/test/logging/integration_test/samples/logger_disablereleasemode.dart:4]: I thought someone else change it cause It shows compile error on my side, Let me double check

**Review by FilledStacks — APPROVED** [review]
LGTM!

**Pull request merged**